### PR TITLE
Make copy paste data agnostic

### PR DESCRIPTION
**Pull request by TrySound**
Copy paste of instances was implemented closly coupled to our schema. Here I make I implement it with simpler interface without dependency on types. Type and version are instead passed as configuration.

Next step will be decoupling instance copy paste from react and simplifying data management with nanostores.

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - test it on preview
- [ ] hi @rpominov, I need you to do
  - detailed review (read every line)

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio-designer/blob/main/apps/designer/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `designer/env-check.js` if mandatory

### CODE REVIEW

**Comment by kof** [comment]
> Copy paste of instances was implemented closly coupled to our schema. Here I make I implement it with simpler interface without dependency on types. Type and version are instead passed as configuration.

I miss the information "why" we need it, what does it unblock?

**Comment by kof** [comment]
Do I understand correctly that we are executing copy/paste logic on both, designer and canvas? and you are trying to make it react decoupled?

We could do this much easer by simply forwarding the copy/paste event and having the logic in designer only. Basically always trying to just forward events from canvas and that's it.

File: apps/designer/app/shared/copy-paste/copy-paste-instance.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+const InstanceCopyData = z.object({
+  instance: Instance,
+  props: Props,
+  styles: Styles,
+});
+
+export type InstanceCopyData = z.infer<typeof InstanceCopyData>;
+
+type InstanceCopyPasteProps = {
+  selectedInstanceData?: InstanceCopyData | undefined;
```
Comment (kof) [inline apps/designer/app/shared/copy-paste/copy-paste-instance.ts:17]: I don't know why we need undefined when its optional already

File: apps/designer/app/shared/copy-paste/copy-paste-instance.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  onCut: (instance: Instance) => void;
+};
+
+// to make it easier to remove React from canvas if we need to
+export const useInstanceCopyPaste = (props: InstanceCopyPasteProps): void => {
+  const { selectedInstanceData, allowAnyTarget, onPaste, onCut } = props;
+  useEffect(() => {
+    selectedInstanceData;
+    onPaste;
+    allowAnyTarget;
```
Comment (kof) [inline apps/designer/app/shared/copy-paste/copy-paste-instance.ts:29]: wow, what is this doing?

File: apps/designer/app/shared/copy-paste/copy-paste-instance.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    return startCopyPaste({
+      version: "@webstudio/instance/v0.1",
```
Comment (kof) [inline apps/designer/app/shared/copy-paste/copy-paste-instance.ts:31]: we should probably get the version somehow from package

File: apps/designer/app/shared/copy-paste/copy-paste-instance.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      type: InstanceCopyData,
+      onCopy: () => {
+        if (selectedInstanceData === undefined) {
+          return;
+        }
+        return selectedInstanceData;
```
Comment (kof) [inline apps/designer/app/shared/copy-paste/copy-paste-instance.ts:37]: this is pointless, can be removed

File: apps/designer/app/shared/copy-paste/copy-paste.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
+import { z } from "zod";
+
+const isValidClipboardEvent = (
+  event: ClipboardEvent,
+  options: { allowAnyTarget: boolean }
+) => {
+  const selection = document.getSelection();
+  if (selection !== null && selection.type === "Range") {
```
Comment (kof) [inline apps/designer/app/shared/copy-paste/copy-paste.ts:8]:
```suggestion
  if (selection?.type === "Range") {
```

File: apps/designer/app/shared/copy-paste/copy-paste-instance.ts (hunk 2 above)
Comment (TrySound) [reply apps/designer/app/shared/copy-paste/copy-paste-instance.ts:29]: Looks like paste works for non inputs with it

Comment (TrySound) [reply apps/designer/app/shared/copy-paste/copy-paste-instance.ts]: Ah, gotcha. Forgot to remove

**Comment by TrySound** [comment]
The main idea is to split all copy/paste setup from data which was spread over few modules.
This will let me easier migrate to style sources.

File: apps/designer/app/shared/copy-paste/copy-paste-instance.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+          return {
+            ...styleDecl,
+            instanceId: instance.id,
+          };
+        });
+
+        onPaste({ instance, props, styles });
+      },
+    });
+  }, [selectedInstanceData, allowAnyTarget, onPaste, onCut]);
```
Comment (rpominov) [inline apps/designer/app/shared/copy-paste/copy-paste-instance.ts:62]: A bit weird that we re-start copy-paste every time `selectedInstanceData` changes. 

Probably not a big deal, but I like the previous version with `start/stop/setProps` a bit more even though it's ugly

**Review by rpominov — APPROVED** [review]
Couple comments with not strong opinions. Up to you what to do with them :)

File: apps/designer/app/shared/copy-paste/copy-paste-instance.ts (hunk 6 above)
Comment (rpominov) [reply apps/designer/app/shared/copy-paste/copy-paste-instance.ts:62]: In general, I think `start/stop/setProps` interface is a good default when you want to make a logic from a hook react-agnostic

Comment (TrySound) [reply apps/designer/app/shared/copy-paste/copy-paste-instance.ts:62]: This is the thing. We don't need react to pass data here. We can access it from stores directly.

File: apps/designer/app/shared/copy-paste/copy-paste.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
+import { z } from "zod";
+
+const isValidClipboardEvent = (
```
[the remark this answers is not in the document]
Comment (TrySound) [reply apps/designer/app/shared/copy-paste/copy-paste.ts:3]: Yeah, let's reconsider this when add new copy paste.

[the remark this answers is not in the document]
Comment (kof) [reply apps/designer/app/shared/copy-paste/copy-paste.ts:3]: future use case can be copy pasting webflow instances, where we would need to differentiate

**Pull request merged**